service/storage_proxy.cc

```diff
@@ -5199,6 +5266,77 @@ void storage_proxy::init_messaging_service(shared_ptr<migration_manager> mm) {
         const bool expired = co_await check_hint_queue_sync_point(request.sync_point_id);
         co_return db::hints::sync_point_check_response{expired};
     });
+
+    ms.register_forward_request([this](query::forward_request req) -> future<query::forward_result> {
+        schema_ptr schema = local_schema_registry().get(req.cmd.schema_version);
+        auto command = make_lw_shared<query::read_command>(std::move(req.cmd));
+
+        slogger.debug("forward_request: received req={} from addr={}", req, _messaging.listen_address());
+
+        // create mocked selection
+        auto selection = [&]{
+            switch (req.type) {
+                case query::forward_request::reduction_type::count: {
+                    auto raw_selectable = cql3::selection::selectable::with_function::raw::make_count_rows_function();
+                    auto column_identifier = make_shared<cql3::column_identifier>("count", false);
+                    std::vector<shared_ptr<cql3::selection::raw_selector>> raw_selectors;
+                    raw_selectors.emplace_back(make_shared<cql3::selection::raw_selector>(raw_selectable, column_identifier));
+                    return cql3::selection::selection::from_selectors(_db.local(), schema, std::move(raw_selectors));
+                }
+                default:
+                    throw std::runtime_error("not implemented reduction type");
+            }
+        }();
+
+        // create pager object
+        auto now = gc_clock::now();
+        auto page_size = req.page_size;
+        auto timeout = req.timeout;
+        auto cql_serialization_format = cql_serialization_format::latest_version;
+        std::vector<size_t> empty_group_by_indices;
+        std::vector<cql3::raw_value> empty_values;
+        std::optional<std::vector<sstring_view>> empty_names;
+        auto query_state = make_lw_shared<service::query_state>(client_state::for_internal_calls(), nullptr,
+                                                                empty_service_permit());
+        auto query_options = make_lw_shared<cql3::query_options>(cql3::default_cql_config, req.cl, empty_names,
+                                                                 empty_values, req.skip_metadata,
+                                                                 cql3::query_options::specific_options::DEFAULT,
+                                                                 cql_serialization_format::latest());
+        auto p = service::pager::query_pagers::pager(schema, selection, *query_state, *query_options, command,
+                                                     std::move(req.pr), nullptr);
```

Comment:
To increase the parallelism even further and reduce the need for moving data between shards, here you could probably split the query range again - this time with a per-shard granularity.

Reply:
I think that would result in an explosion of ranges. Not sure it is worth it, we already have a quite efficient shard-merging in `query_data_from_all_shards()`.

Reply:
Yes.